### PR TITLE
Add huber loss support

### PR DESCRIPTION
**Pull request by jrabary**
This PR adds huber loss support. Used for object detection for instance.

### CODE REVIEW

File: Sources/TensorFlow/Loss.swift [hunk 1]
```diff
@@ -290,3 +290,31 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+///     quadratic to linear. 
+///   - reduction: Reduction to apply on the computed element-wise loss values.
+@differentiable(wrt: predicted)
+public func huberLoss<Scalar: TensorFlowFloatingPoint>(
+    predicted: Tensor<Scalar>,
+    expected: Tensor<Scalar>,
+    delta: Scalar,
+    reduction: @differentiable (Tensor<Scalar>) -> Tensor<Scalar> = { $0.sum() }
+) -> Tensor<Scalar> {
+
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]: Please remove empty line.

File: Sources/TensorFlow/Loss.swift [hunk 2]
```diff
@@ -290,3 +290,31 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+///   - expected: Expected values, i.e. targets, that correspond to the correct output.
+///   - delta: A floating point scalar, the point where the Huber loss function changes from a
+///     quadratic to linear. 
+///   - reduction: Reduction to apply on the computed element-wise loss values.
+@differentiable(wrt: predicted)
+public func huberLoss<Scalar: TensorFlowFloatingPoint>(
+    predicted: Tensor<Scalar>,
+    expected: Tensor<Scalar>,
+    delta: Scalar,
+    reduction: @differentiable (Tensor<Scalar>) -> Tensor<Scalar> = { $0.sum() }
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]: Just a note: some other `reduction: @differentiable (Tensor<Scalar>) -> Tensor<Scalar` default arguments are using a special function `_mean` due to [some linker errors](https://bugs.swift.org/projects/TF/issues/TF-690).

I'm not sure if these linker errors are still reproducible, but the underlying infrastructure regarding these linker issues will be revamped soon.

Comment (jrabary) [reply Sources/TensorFlow/Loss.swift]: @dan-zheng Do you suggest to use _mean for the moment and eventually change it later on ?I'm able to run the test on my computer with the toolchain 0.5 but I don't know if it means something.

File: Sources/TensorFlow/Loss.swift [hunk 3]
```diff
@@ -290,3 +290,30 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     let negAbsLogits = max(logits, -logits) // Custom `abs` to compute gradients at `0`.
     return reduction(maxLogitsWithZero - logits * labels + log1p(exp(-negAbsLogits)))
 }
+
+/// Returns the Huber loss between predictions and expectations.
+///
+/// For each value x in error = exprected - predicted
+///     loss = 0.5 * x^2                 if |x| <= ∂
+///     loss = 0.5 * ∂^2 + ∂ * (|x| - ∂) if |x| > ∂
+/// See: https://en.wikipedia.org/wiki/Huber_loss
```
Comment (eaplatanios) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
/// - Source: [Wikipedia article](https://en.wikipedia.org/wiki/Huber_loss).
```

File: Sources/TensorFlow/Loss.swift [hunk 4]
```diff
@@ -290,3 +290,30 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     let negAbsLogits = max(logits, -logits) // Custom `abs` to compute gradients at `0`.
     return reduction(maxLogitsWithZero - logits * labels + log1p(exp(-negAbsLogits)))
 }
+
+/// Returns the Huber loss between predictions and expectations.
+///
+/// For each value x in error = exprected - predicted
```
Comment (eaplatanios) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
/// For each value `x` in the difference `exprected - predicted`, the loss equals `0.5 * x^2` if
```

File: Sources/TensorFlow/Loss.swift [hunk 5]
```diff
@@ -290,3 +290,30 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+///     loss = 0.5 * x^2                 if |x| <= ∂
```
Comment (eaplatanios) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
/// `abs(x) <= δ`, or `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
```

File: Tests/TensorFlowTests/LossTests.swift [hunk 6]
```diff
@@ -217,6 +217,23 @@ final class LossTests: XCTestCase {
...
 
+    func testHuberLoss() {
+        let predictions = Tensor<Float>([[0.9, 0.2, 0.2], 
+                                         [0.8, 0.4, 0.6]])
+        let targets = Tensor<Float>([[1, 0, 1], 
+                                     [1, 0, 0]])
+
+        // test huber(x, x) = 0
+        let loss1 = huberLoss(predicted: predictions, expected: predictions, delta: Float(1))
+        XCTAssertTrue(loss1.isAlmostEqual(to: Tensor(0), tolerance: 0.001), "Huber Loss all correct")
```
Comment (dan-zheng) [inline Tests/TensorFlowTests/LossTests.swift]: Please use `assertEqual` instead, as other tests are using. This provides more information than `XCTAssertTrue`.

```suggestion
        assertEqual(loss1, Tensor(0), accuracy: 1e-6)
```

Same below for testing `loss2` please.

File: Sources/TensorFlow/Loss.swift [hunk 7]
```diff
@@ -290,3 +290,30 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     let negAbsLogits = max(logits, -logits) // Custom `abs` to compute gradients at `0`.
     return reduction(maxLogitsWithZero - logits * labels + log1p(exp(-negAbsLogits)))
 }
+
+/// Returns the Huber loss between predictions and expectations.
+///
+/// For each value `x` in the difference `exprected - predicted`, the loss equals `0.5 * x^2` if
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
/// For each value `x` in the difference `expected - predicted`, the loss is:
/// - `0.5 * x^2` if `abs(x) <= δ`.
/// - `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
```

File: Sources/TensorFlow/Loss.swift [hunk 8]
```diff
@@ -290,3 +290,30 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+/// `abs(x) <= δ`, or `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
+///
+/// - Source: [Wikipedia article](https://en.wikipedia.org/wiki/Huber_loss).
+///
+/// - Parameters:
+///   - predicted: Predicted outputs from a neural network.
+///   - expected: Expected values, i.e. targets, that correspond to the correct output.
+///   - delta: A floating point scalar, the point where the Huber loss function changes from a
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
///   - delta: A floating point scalar, the point where the Huber loss function changes from
```

File: Tests/TensorFlowTests/LossTests.swift [hunk 9]
```diff
@@ -217,6 +217,36 @@ final class LossTests: XCTestCase {
         assertEqual(computedGradient, expectedGradient, accuracy: 1e-6)
     }
 
+    func testHuberLoss() {
+        let predictions = Tensor<Float>([[0.9, 0.2, 0.2],
```
Comment (dan-zheng) [inline Tests/TensorFlowTests/LossTests.swift]: Some formatting/comment suggestions:
```swift
    func testHuberLoss() {
        let predictions = Tensor<Float>([[0.9, 0.2, 0.2], [0.8, 0.4, 0.6]])
        let labels = Tensor<Float>([[1, 0, 1], [1, 0, 0]])

        do {
          // Test adapted from:
          // https://github.com/tensorflow/tensorflow/blob/148f07323f97ef54998f28cd95c195064ce2c426/tensorflow/python/keras/losses_test.py#L1554
          let loss = huberLoss(predicted: predictions, expected: predictions, delta: 1)
          assertEqual(loss, Tensor(0), accuracy: 1e-6)
        }

        do {
          // Test adapted from:
          // https://github.com/tensorflow/tensorflow/blob/148f07323f97ef54998f28cd95c195064ce2c426/tensorflow/python/keras/losses_test.py#L1560
          // The expected loss was computed using Python TensorFlow 2.0.0-beta1:
          // ```
          // import tensorflow as tf # 2.0.0-beta1
          // predictions = tf.constant([[0.9, 0.2, 0.2], [0.8, 0.4, 0.6]])
          // labels = tf.constant([[1.0, 0.0, 1.0], [1.0, 0.0, 0.0]])
          // loss = tf.losses.Huber(delta=1.0, reduction=tf.losses.Reduction.SUM)
          // print(loss(labels, predictions))
          // # tf.Tensor(0.62500006, shape=(), dtype=float32)
          // ```
          let loss = huberLoss(predicted: predictions, expected: labels, delta: Float(1))
          assertEqual(loss, Tensor(0.62500006), accuracy: 1e-6)
        }
    }
```

File: Tests/TensorFlowTests/LossTests.swift [hunk 10]
```diff
@@ -216,6 +216,34 @@ final class LossTests: XCTestCase {
             [-0.0625, -0.01490036, 0.04759964, 0.0]])
         assertEqual(computedGradient, expectedGradient, accuracy: 1e-6)
     }
+
```
Comment (dan-zheng) [inline Tests/TensorFlowTests/LossTests.swift]: Minor: please remove whitespace-only lines.

```suggestion

```

If you use Xcode on macOS, there's a setting to automate this: `Xcode > Preferences > Text Editing > ☑️ Automatically trim trailing whitespace ☑️ Including whitespace-only lines`.

File: Sources/TensorFlow/Loss.swift [hunk 11]
```diff
@@ -290,3 +290,32 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     let negAbsLogits = max(logits, -logits) // Custom `abs` to compute gradients at `0`.
     return reduction(maxLogitsWithZero - logits * labels + log1p(exp(-negAbsLogits)))
 }
+
+/// Returns the Huber loss between predictions and expectations.
+///
+/// For each value `x` in the difference `expected - predicted`, the loss is:
+/// - `0.5 * x^2` if `abs(x) <= δ`.
+/// - `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
+/// `abs(x) <= δ`, or `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]: Please delete duplicated line.

```suggestion

```

File: Sources/TensorFlow/Loss.swift [hunk 12]
```diff
@@ -290,3 +290,32 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+///
+/// - Source: [Wikipedia article](https://en.wikipedia.org/wiki/Huber_loss).
+///
+/// - Parameters:
+///   - predicted: Predicted outputs from a neural network.
+///   - expected: Expected values, i.e. targets, that correspond to the correct output.
+///   - delta: A floating point scalar, the point where the Huber loss function changes from
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]:
```suggestion
///   - delta: A floating point scalar representing the point where the Huber loss function changes
/// from quadratic to linear.
```

File: Sources/TensorFlow/Loss.swift [hunk 13]
```diff
@@ -290,3 +290,32 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+/// - `0.5 * x^2` if `abs(x) <= δ`.
+/// - `0.5 * δ^2 + δ * (|x| - δ)` otherwise.
+///
+/// - Source: [Wikipedia article](https://en.wikipedia.org/wiki/Huber_loss).
+///
+/// - Parameters:
+///   - predicted: Predicted outputs from a neural network.
+///   - expected: Expected values, i.e. targets, that correspond to the correct output.
+///   - delta: A floating point scalar representing the point where the Huber loss function changes
+/// from quadratic to linear.
```
Comment (dan-zheng) [inline Sources/TensorFlow/Loss.swift]: Please delete duplicate line. (I don't think GitHub supports multi-line suggestions, so committing my suggestions directly may not work.)
```suggestion
```

Comment (jrabary) [reply Sources/TensorFlow/Loss.swift]: I've just commit the fix. I hope there's no more typos I missed :-)

**Comment by dan-zheng** [comment]
CI failed:
```
/swift-apis/Sources/TensorFlow/StdlibExtensions.swift:217:1: error: type 'Element' does not conform to protocol 'Differentiable'
extension Array.DifferentiableView: VectorProtocol where Element: VectorProtocol {
^
/swift-apis/Sources/TensorFlow/StdlibExtensions.swift:217:1: error: 'VectorProtocol' requires that 'Element' conform to 'Differentiable'
extension Array.DifferentiableView: VectorProtocol where Element: VectorProtocol {
^
/swift-apis/Sources/TensorFlow/StdlibExtensions.swift:217:1: note: requirement specified as 'Element' : 'Differentiable'
extension Array.DifferentiableView: VectorProtocol where Element: VectorProtocol {
^
/swift-apis/Sources/TensorFlow/StdlibExtensions.swift:217:1: note: requirement from conditional conformance of 'Array<Element>.DifferentiableView' to 'AdditiveArithmetic'
extension Array.DifferentiableView: VectorProtocol where Element: VectorProtocol {
...
```

Could you please trying merging from `master`? I believe that should fix the errors.
